src/contents/unrrugable-ai-agents.mdx

```diff
@@ -0,0 +1,219 @@
+        calldata,
+    )
+# print(res)
+
+calldata_hex = "0x" + ''.join(f"{byte:02x}" for byte in res)
+print(calldata_hex)
+```
+
+EZKL generates a ZK proofs as raw calldata sent to the Verifier contract. This means that the proof must be sent using a low level evm `CALL` function instead of how you normally call another smart contract in solidity, more about this in Step 4.
+
+## Step 3. Launch the Tip Token
+
+Deploy a simple ERC20 token that the AI agent will be able to interact with. Deploy the following contract that will mint some tokens to you.
+
+```js
+// SPDX-License-Identifier: MIT
+// Compatible with OpenZeppelin Contracts ^5.0.0
+pragma solidity ^0.8.22;
+
+import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+contract MyToken is ERC20 {
+    constructor() ERC20("Tip Token", "TIPT") {
+        _mint(msg.sender, 21_000_000 ether);
+    }
+}
+```
+
+## Step 4: Build the Tipper Contract
+
+The Tipper contract ensures that tokens are distributed only when a valid ZK proof is submitted. Deploy the following contract by replacing the `HALO2_VERIFIER` and `TIP_TOKEN` with the contract addresses you just deployed. Then, transfer at least `1000` Tip Tokens to this contract.
+
+```js
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity >=0.7.0 <0.9.0;
+
+interface Halo2Verifier {
+    function verifyProof(bytes calldata proof, uint256[] calldata instances) external view;
+}
+
+interface IERC20 {
+    function transfer(address to, uint256 value) external returns (bool);
+}
+
+
+contract AiAgentGuardrail {
+    address HALO2_VERIFIER = 0x84fBBc680F4aB4240e8cF6C488aEca03bfF91d2E;
+    address TIP_TOKEN = 0x84fBBc680F4aB4240e8cF6C488aEca03bfF91d2E;
+
+    function verifyAndSendTip(bytes memory proofCalldata) public {
+        (bool success, bytes memory data) = HALO2_VERIFIER.call(proofCalldata);
+        if(success && data.length == 32 && uint8(data[31]) == 1)
+            IERC20(TIP_TOKEN).transfer(extractTipRecipientFromProof(proofCalldata), 1_000 ether);
+        else 
+            revert("Could not verify proof");
+    }
+
+    function extractTipRecipientFromProof(bytes memory proofCalldata) internal pure returns (address) {
+        bytes memory lastBytes = new bytes(20);
+        for (uint i = 0; i < 20; i++) {
+            lastBytes[20-1-i] = proofCalldata[proofCalldata.length - (1+i*32)];
+        }
+        return address(bytes20(abi.encodePacked(lastBytes)));
+    }
+}
+```
+
+Now call `verifyAndSendTip` by submitting the proof as parameter in the calldata format and once the transaction is confirmed `1000` tokens will be awarded to the tip recipient.
+
+## Next steps
+
+By integrating ZK technology, we can create secure, autonomous AI agents that operate transparently on-chain. Next steps involve deepening into ZK learning and also training an AI model capable of performing interesting and exciting tasks. To get started with ZK I recommend TODO and to get started with training ML models, I recommend TODO.
```

Comment:
Are the "TODO"s here a placeholder for recommended resources?

Reply:
Yes, looking for resources to point to, please LMK if you find something that would fit here

Reply:
What do you think about referencing the [ZK Speedrun tutorial](https://dev.to/turupawn/zk-speedrun-3-dsls-in-15-minutes-noir-circom-zokrates-g3d) you had built? Could also tackle an additional win by migrating this resource to Level Up so we can have additional native articles cross-linking one another.

Haven't really tested any educational materials related to training ML models but the one which come to mind is from [Kaggle](https://www.kaggle.com/learn/intro-to-machine-learning)

Reply:
I linked to LevelUp content and Kaggle. Thanks for the suggestions! And yes, sounds good to migrate the speedrun tutorial :+1: will do that later.

Btw it would be nice to have something like this link (notice #zk at the end)
https://www.levelup.xyz/content#zk
That selects the category automatically when the page is opened

Reply:
Good idea! We should be able to create a ticket as a suggestion for Product team to look into in following PR!